cpp-ch/local-engine/Operator/ScalableMergingAggregatedStep.cpp

```diff
@@ -0,0 +1,393 @@
+{
+    auto transform_params = std::make_shared<DB::AggregatingTransformParams>(pipeline.getHeader(), params, true);
+    pipeline.resize(1);
+    auto build_transform = [&](DB::OutputPortRawPtrs outputs)
+    {
+        DB::Processors new_processors;
+        for (auto & output : outputs)
+        {
+            auto op = std::make_shared<ScalableMergingAggregatedTransform>(pipeline.getHeader(), transform_params, context);
+            new_processors.push_back(op);
+            DB::connect(*output, op->getInputs().front());
+        }
+        return new_processors;
+    };
+    pipeline.transform(build_transform);
+}
+
+void ScalableMergingAggregatedStep::describeActions(DB::IQueryPlanStep::FormatSettings & settings) const
+{
+    return params.explain(settings.out, settings.offset);
+}
+
+void ScalableMergingAggregatedStep::describeActions(DB::JSONBuilder::JSONMap & map) const
+{
+    params.explain(map);
+}
+
+void ScalableMergingAggregatedStep::updateOutputStream()
+{
+    output_stream = createOutputStream(input_streams.front(), params.getHeader(input_streams.front().header, true), getDataStreamTraits());
+}
+
+ScalableMergingAggregatedTransform::ScalableMergingAggregatedTransform(
+    const DB::Block & header_, DB::AggregatingTransformParamsPtr params_, DB::ContextPtr context_)
+    : DB::IProcessor({header_}, {params_->getHeader()})
+    , header(header_)
+    , params(params_)
+    , context(context_)
+    , tmp_data_disk(std::make_unique<DB::TemporaryDataOnDisk>(context_->getTempDataOnDisk()))
+{}
+
+bool ScalableMergingAggregatedTransform::isMemoryOverFlow()
+{
+    #if 1
+    UInt64 current_memory_usage = getMemoryUsage();
+    if (params->params.max_bytes_before_external_group_by && current_memory_usage > params->params.max_bytes_before_external_group_by)
+    {
+        LOG_ERROR(
+            logger,
+            "xxx Memory is overflow. current_memory_usage: {}, max_bytes_before_external_group_by: {}",
+            ReadableSize(current_memory_usage),
+            ReadableSize(params->params.max_bytes_before_external_group_by));
+        return true;
+    }
+    return false;
+    #else
+    if (has_single_level || buckets_data_variants.size() > 1)
+    {
+        // LOG_ERROR(logger, "xxx memory overflow. has_single_level: {}, buckets_data_variants.size(): {}", has_single_level, buckets_data_variants.size());
+        return true;
+    }
+    return false;
+    #endif
+}
+
+void ScalableMergingAggregatedTransform::swithMode()
+{
+    has_two_level = true;
+    mode = OVERFLOW;
+}
+
+size_t ScalableMergingAggregatedTransform::spillBucketDataToDisk(Int32 bucket, DB::Block block)
+{
+    if (!block.rows())
+        return 0;
+    // LOG_ERROR(logger, "xxx spillBucketDataToDisk. bucket: {}, block: {}", bucket, block.dumpStructure());
+    auto * tmp_file = getBucketTempFile(bucket);
+    block.info.bucket_num = bucket;
+    return tmp_file->write(block);
```

Comment:
add some trace log, like file dir, io speed, to trace problems, eg. dist sector error